File: custom-responses-image-generation/scripts/generate-image.mjs
#!/usr/bin/env node

import fs from "node:fs";
import os from "node:os";
import path from "node:path";

const HELP = `
Usage:
  node scripts/generate-image.mjs --prompt "..." --out outputs/image.png [options]

Required:
  --prompt <text>             Image prompt. Use --prompt-file or stdin as alternatives.

Output:
  --out <path>                Output image path. Default: generated.png

Codex config:
  --codex-home <path>         Defaults to $CODEX_HOME or <home>/.codex on Linux, macOS, and Windows
  --base-url <url>            Explicit override. Defaults to Codex provider base_url.
  --response-model <model>    Explicit override. Defaults to Codex top-level model.
  --api-key-env <name>        Environment variable fallback for the API key. Default: OPENAI_API_KEY.

Image generation options:
  --action <generate|edit|auto>
  --image <path>              Input image. Can be repeated.
  --mask <path>               Optional inpainting mask image.
  --image-model <model>
  --size <size>
  --quality <low|medium|high|auto>
  --format <png|webp|jpeg>
  --background <transparent|opaque|auto>
  --input-fidelity <high|low>
  --moderation <auto|low>
  --output-compression <0-100>
  --partial-images <0-3>

Other:
  --dry-run                   Print redacted config and request body without calling the API.
  --json                      Print machine-readable result summary.
  --no-progress               Disable progress messages on stderr while waiting for the API.
  --help

Runtime:
  Requires Node.js 18+ only. No npm packages are needed.
`;

function die(message, code = 1) {
  console.error(`Error: ${message}`);
  process.exit(code);
}

function progress(args, message) {
  if (!args["no-progress"]) {
    console.error(`[image-generation] ${message}`);
  }
}

function startProgress(args) {
  if (args["no-progress"]) return () => {};

  const startedAt = Date.now();
  progress(args, "Request sent. Image generation can take several minutes; wait for this command to finish before retrying.");
  const timer = setInterval(() => {
    const elapsedSeconds = Math.round((Date.now() - startedAt) / 1000);
    progress(args, `Still waiting for image result (${elapsedSeconds}s elapsed). Do not start another generation for the same request unless this command fails.`);
  }, 15000);
  if (typeof timer.unref === "function") timer.unref();

  return () => clearInterval(timer);
}

function parseArgs(argv) {
  const args = { image: [] };
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith("--")) {
      die(`unexpected argument: ${token}`);
    }

    const key = token.slice(2);
    if (key === "api-key") {
      die("--api-key was removed to avoid exposing secrets in command lines. Use Codex auth.json or --api-key-env <name>.");
    }
    if (["help", "dry-run", "json", "no-progress"].includes(key)) {
      args[key] = true;
      continue;
    }

    const value = argv[i + 1];
    if (!value || value.startsWith("--")) {
      die(`missing value for --${key}`);
    }
    i += 1;

    if (key === "image") {
      args.image.push(value);
    } else {
      args[key] = value;
    }
  }
  return args;
}

function parseTomlValue(raw) {
  const value = raw.trim();
  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
    return value.slice(1, -1);
  }
  if (value === "true") return true;
  if (value === "false") return false;
  if (/^-?\d+(\.\d+)?$/.test(value)) return Number(value);
  return value;
}

function parseTomlLite(text) {
  const root = {};
  const sections = {};
  let current = root;

  for (const line of text.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;

    const sectionMatch = trimmed.match(/^\[([^\]]+)\]$/);
    if (sectionMatch) {
      const section = sectionMatch[1].replaceAll('"', "");
      sections[section] = sections[section] || {};
      current = sections[section];
      continue;
    }

    const keyValueMatch = trimmed.match(/^([A-Za-z0-9_.-]+)\s*=\s*(.+)$/);
    if (!keyValueMatch) continue;
    const [, key, rawValue] = keyValueMatch;
    current[key] = parseTomlValue(rawValue);
  }

  return { root, sections };
}

function readJsonIfExists(filePath) {
  if (!fs.existsSync(filePath)) return {};
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch (error) {
    die(`failed to parse ${filePath}: ${error.message}`);
  }
}

function envValue(name) {
  if (process.env[name] !== undefined) return process.env[name];
  const lowerName = name.toLowerCase();
  for (const [key, value] of Object.entries(process.env)) {
    if (key.toLowerCase() === lowerName) return value;
  }
  return undefined;
}

function validateEnvName(name, optionName) {
  if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(name)) {
    die(`${optionName} must be an environment variable name, not a secret value.`);
  }
}

function defaultCodexHome() {
  return path.resolve(envValue("CODEX_HOME") || path.join(os.homedir(), ".codex"));
}

function resolveCodexConfig(args) {
  const codexHome = args["codex-home"] ? path.resolve(args["codex-home"]) : defaultCodexHome();
  const configPath = path.join(codexHome, "config.toml");
  const authPath = path.join(codexHome, "auth.json");

  let codexConfig = { root: {}, sections: {} };
  if (fs.existsSync(configPath)) {
    codexConfig = parseTomlLite(fs.readFileSync(configPath, "utf8"));
  }

  const providerName = codexConfig.root.model_provider || "OpenAI";
  const provider = codexConfig.sections[`model_providers.${providerName}`] || {};
  const auth = readJsonIfExists(authPath);
  const apiKeyEnvName = args["api-key-env"] || "OPENAI_API_KEY";
  validateEnvName(apiKeyEnvName, "--api-key-env");

  const baseUrl = args["base-url"] || provider.base_url || envValue("OPENAI_BASE_URL") || "https://api.openai.com/v1";
  const responseModel = args["response-model"] || codexConfig.root.model || envValue("OPENAI_MODEL");
  const authApiKey = auth.OPENAI_API_KEY;
  const envApiKey = envValue(apiKeyEnvName);
  const apiKey = authApiKey || envApiKey;
  const apiKeySource = authApiKey ? "codex-auth" : envApiKey ? `env:${apiKeyEnvName}` : "none";

  if (!responseModel) {
    die("no Responses model found. Set Codex top-level model or pass --response-model.");
  }
  if (!apiKey && !args["dry-run"]) {
    die(`no API key found. Expected OPENAI_API_KEY in Codex auth.json or environment variable ${apiKeyEnvName}.`);
  }

  return {
    codexHome,
    configPath,
    authPath,
    providerName,
    provider,
    baseUrl: baseUrl.replace(/\/+$/, ""),
    responseModel,
    apiKey,
    apiKeySource,
    hasApiKey: Boolean(apiKey),
  };
}

async function readPrompt(args) {
  if (args.prompt) return args.prompt;
  if (args["prompt-file"]) return fs.readFileSync(args["prompt-file"], "utf8").trim();
  if (!process.stdin.isTTY) {
    const chunks = [];
    for await (const chunk of process.stdin) chunks.push(chunk);
    const stdinPrompt = Buffer.concat(chunks).toString("utf8").trim();
    if (stdinPrompt) return stdinPrompt;
  }
  die("missing --prompt, --prompt-file, or stdin prompt.");
}

function mimeTypeFor(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === ".jpg" || ext === ".jpeg") return "image/jpeg";
  if (ext === ".webp") return "image/webp";
  if (ext === ".gif") return "image/gif";
  return "image/png";
}

function imageFileToDataUrl(filePath) {
  if (!fs.existsSync(filePath)) die(`image file not found: ${filePath}`);
  const data = fs.readFileSync(filePath).toString("base64");
  return `data:${mimeTypeFor(filePath)};base64,${data}`;
}

function buildTool(args) {
  const tool = { type: "image_generation" };
  const optionMap = {
    action: "action",
    background: "background",
    "input-fidelity": "input_fidelity",
    "image-model": "model",
    moderation: "moderation",
    "output-compression": "output_compression",
    format: "output_format",
    "partial-images": "partial_images",
    quality: "quality",
    size: "size",
  };

  for (const [argName, bodyName] of Object.entries(optionMap)) {
    if (args[argName] === undefined) continue;
    if (["output-compression", "partial-images"].includes(argName)) {
      const numberValue = Number(args[argName]);
      if (!Number.isFinite(numberValue)) die(`--${argName} must be a number`);
      tool[bodyName] = numberValue;
    } else {
      tool[bodyName] = args[argName];
    }
  }

  if (args.mask) {
    tool.input_image_mask = { image_url: imageFileToDataUrl(args.mask) };
  }

  if (!tool.action && (!args.image || args.image.length === 0)) {
    tool.action = "generate";
  }

  return tool;
}

function buildInput(prompt, args) {
  if (!args.image || args.image.length === 0) return prompt;

  return [
    {
      role: "user",
      content: [
        { type: "input_text", text: prompt },
        ...args.image.map((imagePath) => ({
          type: "input_image",
          image_url: imageFileToDataUrl(imagePath),
        })),
      ],
    },
  ];
}

function outputPathFor(basePath, index, count, format) {
  if (count === 1) return basePath;
  const parsed = path.parse(basePath);
  const ext = parsed.ext || `.${format || "png"}`;
  return path.join(parsed.dir, `${parsed.name}-${index + 1}${ext}`);
}

function redactRequest(body) {
  return JSON.parse(
    JSON.stringify(body, (key, value) => {
      if (key === "image_url" && typeof value === "string" && value.startsWith("data:")) {
        return value.slice(0, value.indexOf(",") + 1) + "<base64-redacted>";
      }
      return value;
    }),
  );
}

function summarizeResponse(responseJson) {
  return {
    id: responseJson.id,
    status: responseJson.status,
    error: responseJson.error?.message || responseJson.error,
    output: (responseJson.output || []).map((item) => ({
      type: item.type,
      status: item.status,
      role: item.role,
      content_types: Array.isArray(item.content) ? item.content.map((content) => content.type) : undefined,
      error: item.error?.message || item.error,
    })),
  };
}

async function main() {
  if (typeof fetch !== "function") {
    die("Node.js 18+ is required because this script uses the built-in fetch API.");
  }

  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    console.log(HELP.trim());
    return;
  }

  const prompt = await readPrompt(args);
  const config = resolveCodexConfig(args);
  const outputPath = args.out || "generated.png";
  const tool = buildTool(args);
  const requestBody = {
    model: config.responseModel,
    input: buildInput(prompt, args),
    tools: [tool],
  };

  if (args["dry-run"]) {
    console.log(JSON.stringify({
      codex_home: config.codexHome,
      config_path: config.configPath,
      auth_path: config.authPath,
      provider: config.providerName,
      base_url: config.baseUrl,
      endpoint: `${config.baseUrl}/responses`,
      response_model: config.responseModel,
      has_api_key: config.hasApiKey,
      api_key_source: config.apiKeySource,
      request: redactRequest(requestBody),
    }, null, 2));
    return;
  }

  let response;
  let responseText;
  const stopProgress = startProgress(args);
  try {
    response = await fetch(`${config.baseUrl}/responses`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${config.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(requestBody),
    });

    responseText = await response.text();
  } finally {
    stopProgress();
  }
  progress(args, "Response received. Decoding image data.");

  let responseJson;
  try {
    responseJson = JSON.parse(responseText);
  } catch {
    die(`API returned non-JSON response with status ${response.status}: ${responseText.slice(0, 500)}`);
  }

  if (!response.ok) {
    const message = responseJson.error?.message || responseJson.message || JSON.stringify(responseJson).slice(0, 1000);
    die(`API request failed with status ${response.status}: ${message}`);
  }

  const imageResults = [];
  for (const item of responseJson.output || []) {
    if (item.type === "image_generation_call" && item.result) {
      imageResults.push(item.result);
    }
  }

  if (imageResults.length === 0) {
    console.error(JSON.stringify(summarizeResponse(responseJson), null, 2));
    die("response did not contain output[].type == image_generation_call with a result.");
  }

  const outputFormat = tool.output_format || path.extname(outputPath).replace(".", "") || "png";
  const written = [];
  for (let i = 0; i < imageResults.length; i += 1) {
    const target = outputPathFor(outputPath, i, imageResults.length, outputFormat);
    fs.mkdirSync(path.dirname(path.resolve(target)), { recursive: true });
    fs.writeFileSync(target, Buffer.from(imageResults[i], "base64"));
    written.push(target);
  }

  const summary = {
    response_id: responseJson.id,
    provider: config.providerName,
    base_url: config.baseUrl,
    response_model: config.responseModel,
    image_model: tool.model || "api-default",
    outputs: written,
  };

  if (args.json) {
    console.log(JSON.stringify(summary, null, 2));
  } else {
    for (const filePath of written) {
      console.log(`Wrote ${filePath}`);
    }
  }
}

main().catch((error) => {
  die(error.stack || error.message || String(error));
});

File: custom-responses-image-generation/scripts/generate-image.py
#!/usr/bin/env python3

import argparse
import base64
import json
import mimetypes
import os
from pathlib import Path
import re
import sys
import urllib.error
import urllib.request


HELP = """Generate or edit images with OpenAI Responses API image_generation.

Runtime:
  Requires Python 3 only. No pip packages are needed.
"""


def die(message, code=1):
    print(f"Error: {message}", file=sys.stderr)
    raise SystemExit(code)


def parse_args():
    parser = argparse.ArgumentParser(description=HELP)
    parser.add_argument("--prompt")
    parser.add_argument("--prompt-file")
    parser.add_argument("--out", default="generated.png")
    parser.add_argument("--codex-home")
    parser.add_argument("--base-url")
    parser.add_argument("--response-model")
    parser.add_argument("--api-key", dest="deprecated_api_key", help=argparse.SUPPRESS)
    parser.add_argument("--api-key-env")
    parser.add_argument("--action", choices=["generate", "edit", "auto"])
    parser.add_argument("--image", action="append", default=[])
    parser.add_argument("--mask")
    parser.add_argument("--image-model")
    parser.add_argument("--size")
    parser.add_argument("--quality", choices=["low", "medium", "high", "auto"])
    parser.add_argument("--format", choices=["png", "webp", "jpeg"])
    parser.add_argument("--background", choices=["transparent", "opaque", "auto"])
    parser.add_argument("--input-fidelity", choices=["high", "low"])
    parser.add_argument("--moderation", choices=["auto", "low"])
    parser.add_argument("--output-compression", type=int)
    parser.add_argument("--partial-images", type=int)
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()
    if args.deprecated_api_key is not None:
        die("--api-key was removed to avoid exposing secrets in command lines. Use Codex auth.json or --api-key-env <name>.")
    return args


def parse_toml_value(raw):
    value = raw.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in ("'", '"'):
        return value[1:-1]
    if value == "true":
        return True
    if value == "false":
        return False
    if re.match(r"^-?\d+(\.\d+)?$", value):
        return float(value) if "." in value else int(value)
    return value


def parse_toml_lite(text):
    root = {}
    sections = {}
    current = root

    for line in text.splitlines():
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue

        section_match = re.match(r"^\[([^\]]+)\]$", trimmed)
        if section_match:
            section = section_match.group(1).replace('"', "")
            current = sections.setdefault(section, {})
            continue

        key_value_match = re.match(r"^([A-Za-z0-9_.-]+)\s*=\s*(.+)$", trimmed)
        if not key_value_match:
            continue
        key, raw_value = key_value_match.groups()
        current[key] = parse_toml_value(raw_value)

    return {"root": root, "sections": sections}


def read_json_if_exists(file_path):
    if not file_path.exists():
        return {}
    try:
        return json.loads(file_path.read_text(encoding="utf-8"))
    except Exception as error:
        die(f"failed to parse {file_path}: {error}")


def env_value(name):
    if name in os.environ:
        return os.environ[name]
    lower_name = name.lower()
    for key, value in os.environ.items():
        if key.lower() == lower_name:
            return value
    return None


def validate_env_name(name, option_name):
    if not re.match(r"^[A-Za-z_][A-Za-z0-9_]*$", name):
        die(f"{option_name} must be an environment variable name, not a secret value.")


def default_codex_home():
    return Path(env_value("CODEX_HOME") or Path.home() / ".codex").resolve()


def resolve_codex_config(args):
    codex_home = Path(args.codex_home).resolve() if args.codex_home else default_codex_home()
    config_path = codex_home / "config.toml"
    auth_path = codex_home / "auth.json"

    codex_config = {"root": {}, "sections": {}}
    if config_path.exists():
        codex_config = parse_toml_lite(config_path.read_text(encoding="utf-8"))

    provider_name = codex_config["root"].get("model_provider") or "OpenAI"
    provider = codex_config["sections"].get(f"model_providers.{provider_name}", {})
    auth = read_json_if_exists(auth_path)
    api_key_env_name = args.api_key_env or "OPENAI_API_KEY"
    validate_env_name(api_key_env_name, "--api-key-env")

    base_url = (
        args.base_url
        or provider.get("base_url")
        or env_value("OPENAI_BASE_URL")
        or "https://api.openai.com/v1"
    ).rstrip("/")
    response_model = args.response_model or codex_config["root"].get("model") or env_value("OPENAI_MODEL")
    auth_api_key = auth.get("OPENAI_API_KEY")
    env_api_key = env_value(api_key_env_name)
    api_key = auth_api_key or env_api_key
    api_key_source = "codex-auth" if auth_api_key else f"env:{api_key_env_name}" if env_api_key else "none"

    if not response_model:
        die("no Responses model found. Set Codex top-level model or pass --response-model.")
    if not api_key and not args.dry_run:
        die(f"no API key found. Expected OPENAI_API_KEY in Codex auth.json or environment variable {api_key_env_name}.")

    return {
        "codex_home": str(codex_home),
        "config_path": str(config_path),
        "auth_path": str(auth_path),
        "provider_name": provider_name,
        "base_url": base_url,
        "response_model": response_model,
        "api_key": api_key,
        "api_key_source": api_key_source,
        "has_api_key": bool(api_key),
    }


def read_prompt(args):
    if args.prompt:
        return args.prompt
    if args.prompt_file:
        return Path(args.prompt_file).read_text(encoding="utf-8").strip()
    if not sys.stdin.isatty():
        prompt = sys.stdin.read().strip()
        if prompt:
            return prompt
    die("missing --prompt, --prompt-file, or stdin prompt.")


def mime_type_for(file_path):
    guessed, _ = mimetypes.guess_type(str(file_path))
    return guessed or "image/png"


def image_file_to_data_url(file_path):
    path = Path(file_path)
    if not path.exists():
        die(f"image file not found: {file_path}")
    data = base64.b64encode(path.read_bytes()).decode("ascii")
    return f"data:{mime_type_for(path)};base64,{data}"


def build_tool(args):
    tool = {"type": "image_generation"}
    option_map = {
        "action": "action",
        "background": "background",
        "input_fidelity": "input_fidelity",
        "image_model": "model",
        "moderation": "moderation",
        "output_compression": "output_compression",
        "format": "output_format",
        "partial_images": "partial_images",
        "quality": "quality",
        "size": "size",
    }

    for arg_name, body_name in option_map.items():
        value = getattr(args, arg_name)
        if value is not None:
            tool[body_name] = value

    if args.mask:
        tool["input_image_mask"] = {"image_url": image_file_to_data_url(args.mask)}

    if "action" not in tool and not args.image:
        tool["action"] = "generate"

    return tool


def build_input(prompt, args):
    if not args.image:
        return prompt

    return [
        {
            "role": "user",
            "content": [
                {"type": "input_text", "text": prompt},
                *[
                    {"type": "input_image", "image_url": image_file_to_data_url(image_path)}
                    for image_path in args.image
                ],
            ],
        }
    ]


def output_path_for(base_path, index, count, output_format):
    path = Path(base_path)
    if count == 1:
        return path
    suffix = path.suffix or f".{output_format or 'png'}"
    return path.with_name(f"{path.stem}-{index + 1}{suffix}")


def redact_request(body):
    def redact(value):
        if isinstance(value, dict):
            return {key: redact(item) for key, item in value.items()}
        if isinstance(value, list):
            return [redact(item) for item in value]
        if isinstance(value, str) and value.startswith("data:") and "," in value:
            return value.split(",", 1)[0] + ",<base64-redacted>"
        return value

    return redact(body)


def summarize_response(response_json):
    return {
        "id": response_json.get("id"),
        "status": response_json.get("status"),
        "error": response_json.get("error", {}).get("message") if isinstance(response_json.get("error"), dict) else response_json.get("error"),
        "output": [
            {
                "type": item.get("type"),
                "status": item.get("status"),
                "role": item.get("role"),
                "content_types": [content.get("type") for content in item.get("content", [])] if isinstance(item.get("content"), list) else None,
                "error": item.get("error", {}).get("message") if isinstance(item.get("error"), dict) else item.get("error"),
            }
            for item in response_json.get("output", [])
        ],
    }


def parse_response_json(status, response_bytes):
    text = response_bytes.decode("utf-8", errors="replace")
    try:
        return json.loads(text)
    except Exception:
        die(f"API returned non-JSON response with status {status}: {text[:500]}")


def post_json(endpoint, api_key, body):
    data = json.dumps(body).encode("utf-8")
    request = urllib.request.Request(
        endpoint,
        data=data,
        method="POST",
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(request) as response:
            return response.status, response.read()
    except urllib.error.HTTPError as error:
        return error.code, error.read()


def main():
    args = parse_args()
    prompt = read_prompt(args)
    config = resolve_codex_config(args)
    tool = build_tool(args)
    request_body = {
        "model": config["response_model"],
        "input": build_input(prompt, args),
        "tools": [tool],
    }
    endpoint = f"{config['base_url']}/responses"

    if args.dry_run:
        print(json.dumps({
            "codex_home": config["codex_home"],
            "config_path": config["config_path"],
            "auth_path": config["auth_path"],
            "provider": config["provider_name"],
            "base_url": config["base_url"],
            "endpoint": endpoint,
            "response_model": config["response_model"],
            "has_api_key": config["has_api_key"],
            "api_key_source": config["api_key_source"],
            "request": redact_request(request_body),
        }, indent=2))
        return

    status, response_bytes = post_json(endpoint, config["api_key"], request_body)
    response_json = parse_response_json(status, response_bytes)
    if status < 200 or status >= 300:
        message = response_json.get("error", {}).get("message") or response_json.get("message") or json.dumps(response_json)[:1000]
        die(f"API request failed with status {status}: {message}")

    image_results = [
        item["result"]
        for item in response_json.get("output", [])
        if item.get("type") == "image_generation_call" and item.get("result")
    ]
    if not image_results:
        print(json.dumps(summarize_response(response_json), indent=2), file=sys.stderr)
        die("response did not contain output[].type == image_generation_call with a result.")

    output_format = tool.get("output_format") or Path(args.out).suffix.lstrip(".") or "png"
    written = []
    for index, image_base64 in enumerate(image_results):
        target = output_path_for(args.out, index, len(image_results), output_format)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(base64.b64decode(image_base64))
        written.append(str(target))

    summary = {
        "response_id": response_json.get("id"),
        "provider": config["provider_name"],
        "base_url": config["base_url"],
        "response_model": config["response_model"],
        "image_model": tool.get("model") or "api-default",
        "outputs": written,
    }

    if args.json:
        print(json.dumps(summary, indent=2))
    else:
        for file_path in written:
            print(f"Wrote {file_path}")


if __name__ == "__main__":
    main()
